Source/MQTTnet/Internal/AsyncEvent.cs

```diff
@@ -32,25 +48,46 @@ public void AddHandler(Action<TEventArgs> handler)
                 throw new ArgumentNullException(nameof(handler));
             }
 
-            _handlers.Add(new AsyncEventInvocator<TEventArgs>(handler, null));
+            lock (_handlers)
+            {
+                _handlers.Add(new AsyncEventInvocator<TEventArgs>(handler, null));
+
+                HasHandlers = true;
+                _handlersForInvoke = new List<AsyncEventInvocator<TEventArgs>>(_handlers.ToList());
+            }
         }
 
         public async Task InvokeAsync(TEventArgs eventArgs)
         {
-            foreach (var handler in _handlers)
+            if (!HasHandlers)
+            {
+                return;
+            }
+
+            // Adding or removing handlers will produce a new list instance all the time.
+            // So locking here is not required since only the reference to an immutable list
+            // of handlers is used.
+            var handlers = _handlersForInvoke;
+            foreach (var handler in handlers)
             {
                 await handler.InvokeAsync(eventArgs).ConfigureAwait(false);
             }
```

Comment:
Not related to the PR, just as idea:
Would something like be possible and show some perf-benefits?
```c#
[ThreadStatic]
private static List<Task> t_tasksCache;
// ...
var handlers = _handlersForInvoke;
List<Task> tasks = t_taskCache ?? new List<Task>(capacity: handlers .Count);
foreach (var handler in handlers)
{
    tasks.Add(handler.InvokeAsync(eventArgs));
}
await Task.WhenAll(tasks).ConfigureAwait(false);
tasks.Clear();
t_taskCache = tasks;
```
So issue all invokes at once, then wait for them at once instead of individually waiting.
To avoid the allocation of the list to hold the tasks, a thread-static used here -- but there may be better pooling / caching for this.

Reply:
I don't like the idea because events are usually called in serial. There is also a way to pass data across handlers like "IsHandled" properties etc. When calling all at once in separate tasks the business logic may end up in race conditions. If this behavior is wanted users may just start a new Task in the event handler.

Reply:
Makes sense.